### PR TITLE
feat(cli): use abi types in store config

### PR DESCRIPTION
**Pull request by dk1a**
Config now uses strings for both abi and user types
I also added a function to define configs, because otherwise user types can't really be typehinted
When using the function any missing user types will show a typescript error
(to simplify the logic for this I also removed `userTypes`, so `enums` are in config root now)

### CODE REVIEW

File: packages/cli/src/config/index.ts [hunk 1]
```diff
@@ -1,9 +1,18 @@
...
 import { StoreUserConfig, StoreConfig } from "./parseStoreConfig.js";
 import { WorldUserConfig, ResolvedWorldConfig } from "./world/index.js";
 
-export type MUDUserConfig = StoreUserConfig & WorldUserConfig;
+export type MUDUserConfig<EnumNames extends StringForUnion = StringForUnion> = StoreUserConfig<EnumNames> &
+  WorldUserConfig;
 export type MUDConfig = StoreConfig & ResolvedWorldConfig;
 
+/** Type helper for defining MUDUserConfig */
+export function defineMUDUserConfig<EnumNames extends StringForUnion = StringForUnion>(
```
Comment (holic) [inline packages/cli/src/config/index.ts:10]: nit: would be great if we could just call this `mudConfig({...})`

File: packages/cli/src/render-solidity/renderTypesFromConfig.ts [hunk 2]
```diff
@@ -2,9 +2,9 @@ import { StoreConfig } from "../config/parseStoreConfig.js";
 import { renderTypes } from "./renderTypes.js";
 
 export function renderTypesFromConfig(config: StoreConfig) {
-  const enums = Object.keys(config.userTypes.enums).map((name) => ({
+  const enums = Object.keys(config.enums).map((name) => ({
     name,
-    memberNames: config.userTypes.enums[name],
+    memberNames: config.enums[name],
```
Comment (holic) [inline packages/cli/src/render-solidity/renderTypesFromConfig.ts:7]: 😍

File: packages/cli/src/utils/typeUtils.ts [hunk 3]
```diff
@@ -0,0 +1,5 @@
+export type RequireKeys<T extends Record<string, unknown>, P extends string> = T & Required<Pick<T, P>>;
+
+// This allows unions between string literals and `string` without sacrificing autocompletion.
+// Workaround for https://github.com/Microsoft/TypeScript/issues/29729
+export type StringForUnion = string & Record<never, never>;
```
Comment (holic) [inline packages/cli/src/utils/typeUtils.ts:5]: nice!

File: packages/store/mud.config.mts [hunk 4]
```diff
@@ -1,37 +1,34 @@
...
+        a32: "uint32[]",
+        s: "string",
       },
     },
     Vector2: {
       schema: {
-        x: SchemaType.UINT32,
-        y: SchemaType.UINT32,
+        x: "uint32",
+        y: "uint32",
```
Comment (holic) [inline packages/store/mud.config.mts:30]: ohhh this is sooo much better!

**Review by holic — COMMENTED** [review]
just one small nit about simplifying the name the config function as we gear up towards release!

**Review by holic — APPROVED** [review]

File: packages/cli/src/config/parseStoreConfig.ts [hunk 5]
```diff
@@ -151,22 +169,29 @@ export interface StoreUserConfig {
    * The key is the table name (capitalized).
    *
    * The value:
-   *  - `SchemaType | userType` for a single-value table (aka ECS component).
+   *  - abi or user type for a single-value table (aka ECS component).
```
Comment (alvrs) [inline packages/cli/src/config/parseStoreConfig.ts:172]: not part of this PR, but I think we should remove `aka ECS component`. Having a single value is not the distinguishing factor for an ECS component, since components can have multiple (related) values (eg. `PositionComponent = {x: int32, y: int32 }`). The distinguishing factor for ECS components vs regular tables would rather be having a single key vs a composite key.

File: packages/recs/src/v2/abiTypesToRecsTypes.ts [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+  uint16: RecsType.Number,
+  uint24: RecsType.Number,
+  uint32: RecsType.Number,
+  uint40: RecsType.Number,
+  uint48: RecsType.Number,
+  uint56: RecsType.Number,
+  uint64: RecsType.Number,
+  uint72: RecsType.Number,
+  uint80: RecsType.Number,
+  uint88: RecsType.Number,
```
Comment (alvrs) [inline packages/recs/src/v2/abiTypesToRecsTypes.ts:15]: Is anything above `uint53` really represented as `number` in `recs`? Since those don't fit into javascript `number` and `recs` in its current state doesn't support `bigint`, I would expect them to be represented as a hex string (= `RecsType.String`)

File: packages/cli/src/config/index.ts (hunk 1 above)
Comment (alvrs) [reply packages/cli/src/config/index.ts]: For consistency we should probably capitalize `MUD` here.

File: packages/recs/src/v2/abiTypesToRecsTypes.ts (hunk 6 above)
Comment (dk1a) [reply packages/recs/src/v2/abiTypesToRecsTypes.ts:15]: Defer to @holic (this isn't really a part of this PR, as I just refactored the previous mapping of schema types)

File: packages/cli/src/config/index.ts (hunk 1 above)
Comment (dk1a) [reply packages/cli/src/config/index.ts]: should `storeConfig` be `StoreConfig` then to consistently capitalize them too?

File: packages/recs/src/v2/abiTypesToRecsTypes.ts (hunk 6 above)
Comment (holic) [reply packages/recs/src/v2/abiTypesToRecsTypes.ts:15]: does RECS consistently give us back a native number? iirc some things use `BigNumber` under the hood

File: packages/cli/src/config/index.ts (hunk 1 above)
Comment (dk1a) [reply packages/cli/src/config/index.ts]: doing a little refactoring PR rn, capitalized func names don't feel good. And with MUDConfig I also get conflicts with types and zod stuff

Comment (dk1a) [reply packages/cli/src/config/index.ts]: `defineMUDConfig` would be a better alternative, but I'd still prefer `mudConfig` as it's shorter
